### PR TITLE
Add Firebase authentication

### PR DESCRIPTION
**Pull request by amyqxhuang**
## Overview

Add Firebase authentication to allow for more secure database rules

## Test Coverage

Manual testing

### CODE REVIEW

File: Pear/AppDelegate.swift [hunk 1]
```diff
@@ -33,6 +33,12 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         return true
     }
 
+    func application(_ application: UIApplication, open url: URL,
+                     options: [UIApplication.OpenURLOptionsKey: Any])
+      -> Bool {
+        return GIDSignIn.sharedInstance().handle(url)
+    }
+
```
Comment (danielVebman) [inline Pear/AppDelegate.swift:41]:
```suggestion
    func application(
        _ application: UIApplication, 
        open url: URL,
        options: [UIApplication.OpenURLOptionsKey: Any]
    ) -> Bool {
        GIDSignIn.sharedInstance().handle(url)
    }

```

File: Pear/Controller/LoginViewController.swift [hunk 2]
```diff
@@ -132,10 +134,21 @@ extension LoginViewController: GIDSignInDelegate {
               // Only allow users with Cornell emails or internal users (Pear email) to log in
               email.contains("@cornell.edu") || email == "cornellpearapp@gmail.com" else {
             GIDSignIn.sharedInstance().signOut()
+            do {
+                try Auth.auth().signOut()
+            } catch let signOutError as NSError {
```
Comment (danielVebman) [inline Pear/Controller/LoginViewController.swift:139]: Catch clauses implicitly give you an `error`. You can just `} catch {` and print `error` in the body.

File: Pear/Controller/LoginViewController.swift [hunk 3]
```diff
@@ -132,10 +134,21 @@ extension LoginViewController: GIDSignInDelegate {
...
+              print("Error signing out: \(signOutError)")
```
Comment (danielVebman) [inline Pear/Controller/LoginViewController.swift:140]: Can even be 

```suggestion
              print("Error signing out:", signOutError)
```

File: Pear/Settings/SettingsViewController.swift [hunk 4]
```diff
@@ -149,6 +150,11 @@ extension SettingsViewController: UITableViewDataSource {
             pushEditSocialMediaViewController()
         } else if option.text == "Log Out" {
             GIDSignIn.sharedInstance()?.signOut()
+            do {
+                try Auth.auth().signOut()
+            } catch let signOutError as NSError {
+              print("Error signing out: \(signOutError)")
+            }
```
Comment (danielVebman) [inline Pear/Settings/SettingsViewController.swift:157]: Same thing here as the catch above.

File: Pear/Controller/LoginViewController.swift [hunk 5]
```diff
@@ -132,10 +134,21 @@ extension LoginViewController: GIDSignInDelegate {
               // Only allow users with Cornell emails or internal users (Pear email) to log in
               email.contains("@cornell.edu") || email == "cornellpearapp@gmail.com" else {
             GIDSignIn.sharedInstance().signOut()
+            do {
+                try Auth.auth().signOut()
```
Comment (danielVebman) [inline Pear/Controller/LoginViewController.swift:138]: It's been a while, but can you please remind me why we use both `GIDSignIn` and `Auth`?

**Review by danielVebman — APPROVED** [review]
Hi Amy, it's been a while! Left just a couple comments 🤩🚀🚢

File: Pear/Controller/LoginViewController.swift (hunk 5 above)
Comment (amyqxhuang) [reply Pear/Controller/LoginViewController.swift:144]: GIDSignIn doesn't automatically add the user to the authenticated table in Firebase. By having auth, we can check for a user's authenticated status in Firebase database when writing database rules.

**Pull request merged**